composer/utils/checkpoint.py

```diff
@@ -443,23 +453,75 @@ def __init__(self, source_path: str, destination_path: str, object_store):
+                    # The file could have already been downloaded as different plan items can point to same file.
+                    if not os.path.exists(file_destination):
+                        log.debug(f'Downloading {relative_file_path} to {file_destination}.')
+                        self.object_store.download_object(object_name=str(
+                            Path(self.source_path) / Path(relative_file_path)),
+                                                          filename=file_destination)
+                        log.debug(f'Finished downloading {relative_file_path} to {file_destination}.')
 
             # 2. Wait for all ranks to finish.
+            log.debug(f'Finished downloading all files.')
             dist.barrier()
+            log.debug('Done waiting for all ranks to finish downloading files.')
+
+            # 3. Broadcast files to all other replicas if HSDP
+            if device_mesh is not None and device_mesh.ndim == 2:
+                # Broadcast file to all replicas. Assume replica size is at least 1 node
+                replicate_process_group = device_mesh.get_group(0)  # Replicate replicate_process_group
+                shard_size = device_mesh.size(1)
+
+                # Send list of files to all ranks
+                file_list = [sorted(os.listdir(self.destination_path))]
+                dist.broadcast_object_list(file_list,
+                                           src=dist.get_global_rank() % shard_size,
+                                           group=replicate_process_group)
+                file_list = file_list[0]
+                log.debug(f'{file_list=}')
+
+                # Send each file to the appropriate rank
+                for file_name in file_list:
+                    if 'metadata' in file_name:  # All ranks already have the metadata file
+                        continue
+                    if dist.get_local_rank() == 0:
+                        full_path = os.path.join(self.destination_path, file_name)
+                        log.debug(f'Transferring {full_path=}')
+                        file_object = [None]
+                        if dist.get_global_rank() % shard_size == dist.get_global_rank():
+                            # Process with rank 0 reads the file and prepares the object
+                            with open(full_path, 'rb') as f:
+                                file_object = [{'content': f.read()}]
+                            # log.debug(f'md5sum of {full_path=} is {file_object["content"][:10]}')
+                        dist.broadcast_object_list(file_object,
+                                                   src=dist.get_global_rank() % shard_size,
+                                                   group=replicate_process_group)
+                        received_file_object = file_object[0]
+                        assert received_file_object is not None
+                        if dist.get_global_rank() % shard_size != dist.get_global_rank():
```

Comment:
you do `dist.get_global_rank() % shard_size` stuff in a bunch of places. maybe define some descriptive variables for these mods and checks